File: Stage/Python/Core/Stations.py
from Core.Element import C_Element

class C_Stations:

    def __init__(self,xmlFile):
        self._stations = {}
        self._vues = []
        listStations = xmlFile.searchXml("Stations")
        for station in listStations:
            nom = station.get("nom")
            couleur = station.get("couleur")
            
            newStation = C_Element(nom,couleur,"station")
                
            self._stations[nom]= newStation
        

    def abonne(self, vue):
        trouver = 0
        for v in self._vues:
            if (v == vue):
                trouver = 1

        if (trouver==0):
            self._vues.append(vue)

    def getListStation(self):
        listStations = []
        stations = list(sorted(self._stations.keys()))
        for station in stations:
            listStations.append(self._stations.get(station))
            
        return listStations

    def getListStation_Noms(self, listNom):
        listeStations = []
        for nom in listNom:
            listeStations.append(self.getStation(nom))
        return listeStations

    def getStation(self, nom):
        return self._stations.get(nom)

    def getListNomStation(self):
        return list(sorted(self._stations.keys()))

    def getListCouleur(self):
        listCouleur = []
        for station in self._stations.values():
            listCouleur.append(station.getCouleur())
        return listCouleur

    def getCouleurStation(self, nom):
        station = self._stations.get(nom)
        return station.getCouleur()
        
    def setCouleurStation(self, nom , couleur):
        station = self._stations.get(nom)
        station.setCouleur(couleur)
        
        self.update()

    def setNomStation(self, nomAvant, nomChanger):
        station = self._stations.get(nomAvant)
        station.setNom(nomChanger)
        
        self._stations[nomChanger] = self._stations.pop(nomAvant)
        
        self.update()
        
    def update(self):
        for vue in self._vues:
            vue.refresh()

File: Stage/C++/Core/stations.cpp
#include "stations.h"

// classe qui permet de gerer toutes les stations
// herite de la classe modele a chaque modification d'un groupe la classe va appeler la fonction refresh de ces vues(qui sont les interfaces)
// stations est un dictionnaire qui a pour cle le nom de station et comme valeur le pointeur vers la station
/*--------------------------------CONSTRUCTEURS--------------------------------*/
Stations::Stations(Xml *xml){
    QList<QDomElement> listElement = xml->getCible("Stations");

    foreach(QDomElement element, listElement){
        QString nom = element.attribute("nom");
        QString couleur = element.attribute("couleur");
        QString categorie = element.attribute("categorie");
        Station *station = new Station(nom,couleur);
        station->setCategorie(categorie);
        stations.insert(nom,station);
    }
}
/*--------------------------------DESTRUCTEURS--------------------------------*/
Stations::~Stations(){
    foreach(Station* elt,stations.values()){
        if(elt!=NULL)
            delete elt;
    }
    stations.clear();
}

/*--------------------------------FONCTIONS PUBLIC--------------------------------*/
QList<Element*> Stations::getListStation(){
    QList<Element*> listStation;
    foreach(Element* station, stations.values()){
        listStation.append(station);
    }

    return listStation;
}

// fonction qui peremt de recuperer une liste de stations en fonction d'une liste de nom
QList<Element*> Stations::getListStation(QList<QString> listNom){
    QList<Element*> listStation;
    foreach(QString nom, listNom){
        listStation.append(stations.value(nom));
    }

    return listStation;
}

Element* Stations::getStation(QString nom){
    return stations.value(nom);
}

QList<QString> Stations::getListNomStation(){
    return stations.keys();
}

QList<QString> Stations::getListCouleurStation(){
    QList<QString> listCouleur;
    foreach(Station* station , stations.values()){
        listCouleur.append(station->getCouleur());
    }
    return listCouleur;
}

QString Stations::getCouleurStation(QString nom){
    return stations.value(nom)->getCouleur();
}

void Stations::setCouleurStation(QString nom, QString couleur){
    stations.value(nom)->setCouleur(couleur);
    update();
}

void Stations::setNomStation(QString oldNom, QString newNom){
    Station* station = stations.value(oldNom);
    station->setNom(newNom);
    stations.remove(oldNom);
    stations.insert(stations.end(),newNom,station);
    update();
}

// fonction qui permet de savoir si un element est une station
bool Stations::isStation(QString nom){
    return stations.keys().contains(nom);
}

// fonction qui permet d'ajouter un station
void Stations::addStation(QString nom){
    Station *station = new Station(nom,"125,125,125");
    station->setCategorie("Nouveaux");
    stations.insert(nom,station);
    update();
}

void Stations::setCategorie(QString nom,QString categorie){
    Station *station = stations.value(nom);
    station->setCategorie(categorie);
    update();
}

void Stations::removeStation(QString nomStation){
    Station *station = stations.value(nomStation);
    stations.remove(nomStation);
    delete station;
    update();
}
